design-proposals/eim-decouple-microos-storage.md

```diff
@@ -0,0 +1,185 @@
+  It could be a Release Service (S3 bucket) or any other HTTPS file server.
+- The Micro-OS image should be signed and the signing keys (DER keys) should be stored along with the Micro-OS image.
+- The image stored externally is a generic image, without any orchestrator-specific files embedded.
+- The orchestrator (DKAM) will have to provide orchestrator-specific files to ENs. The files package contains orchestrator CA certificate and
+  environment files that are needed for Micro-OS to work with the given Edge Orchestrator instance. The files package will be called a "Micro-OS bundle" hereinafter.
+- The Micro-OS bundle is a signed archive. It is signed by the same keychain as iPXE script.
+- iPXE script along with the Micro-OS bundle are still stored on the orchestrator PV. Since these are small files, the provisioning time would not be impacted.
+- ENs will download the Micro-OS bundle and the Micro-OS image during the iPXE phase and inject the Micro-OS bundle to the Micro-OS image (via `initrd`), similarly to how the initramfs is loaded.
+  For example, the CA certificate may be inserted under `/etc/pki/ca-trust/source/anchors/`, while the environment variables can be injected to `/etc/hook/env_config`.
+- The services running inside Micro-OS (e.g., device discovery, tink worker) should read from these files to retrieve certificate, env variales, and other required files.
+
+The workflow is presented below. Note that this workflow uses an example of Micro-EMT and RS file server to store Micro-EMT.
+The RS file server could be easily replaced by any other HTTPS-based local server that stores Micro-EMT. 
+
+Note that a similar design can be achieved with the current Hook-OS that we can store on the RS file server.
+
+```mermaid
+sequenceDiagram
+%%{wrap}%%
+autonumber
+
+  participant user as User
+
+  participant en as Edge Node
+
+  box rgb(235,255,255) Edge Orchestrator
+
+      participant pa as Provisioning Nginx
+      participant om as Onboarding Manager
+      participant dkam as DKAM
+      participant ic as infra-config
+      participant inv as Inventory / API
+
+  end
+
+  box rgb(235,255,255) Edge Microvisor Toolkit Build System
+      participant emt-ci as EMT CI/CD Pipeline
+  end
+
+  participant rs as Release Service File Server
+
+  rect rgb(191, 223, 255)
+
+  note right of emt-ci: Day -1, Micro-EMT build process
+
+  emt-ci->>emt-ci: Build Micro-EMT with all required agents
+  emt-ci->>emt-ci: Sign generated Micro-EMT with the DER security key
+  emt-ci->>rs: Publish signed Micro-EMT image along with the DER security key 
+
+  end
+
+  ic->>dkam: Read orchestrator CA certificate and cluster-wide variables
+
+  dkam->>dkam: Generates and signs Micro-OS bundle
+
+  dkam->>pa: Store Micro-OS bundle on PV 
+
+  dkam->>dkam: Builds and signs iPXE EFI binary
+
+  dkam->>pa: Store signed iPXE EFI binary
+
+  user->>+rs: Download RS certificate & Micro-OS DER key
```

Comment:
RS Certrificate might not be required use `opt` if the certificte is ss only

Reply:
I'm not sure about this. The signing of Micro-OS is used to add an additional security level when downloading via iPXE, which is only secured by HTTPS with one-way TLS.

Reply:
The iPXE client is not really rich like other clients to carry well known CAs. So, irrespective of the source, we might need to add the trust chain for TLS.